doc/developer/style.md

```diff
@@ -116,6 +116,77 @@ Some key principles to highlight are:
 * Avoid contractions and spell out words in full.
 * Word choices to be mindful of: find vs. exists, may vs. can vs. might.
 
+## Log message style
+
+We use the [`tracing` crate](https://docs.rs/tracing/latest/tracing/)'s log
+macros. Observe the following guidelines:
+
+  * Use sentence fragment style. Say "beginning operation", not
+    "Beginning operation."
+
+  * Prefer structured fields to string formatting:
+
+    ```
+    fn handle_persist_thingy(shard_id: ShardId) {
+        // Preferred.
+        info!(shard_id = ?the_shard, "handling persist thingy");
+        // Acceptable.
+        info!("handling persist thingy for shard {}", shard_id);
+    }
+    ```
+
+Messages at each level must meet the indicated standard:
+
+* **`ERROR`**: Reports an error that certainly indicates data corruption, data
+  loss, unavailability, or an invariant violation that results in undefined
+  behavior. These errors are reported to Sentry and may eventually page the
+  on-call engineer. Use *judiciously*.
+
+  Examples:
+
+  * A referenced collection is not present in the storage controller.
+  * A source ingested the deletion of a record that does not exist, causing a
+    "negative multiplicity."
+
+* **`WARN`**: Like `ERROR`, but when there is some uncertainty about the
```

Comment:
I'd recommend we additionally (or instead?) define ERROR and WARN in terms of the action that should be taken by someone seeing the log. i find that easier to extrapolate that sort of guidance than (necessarily) non-exhaustive categories of "what". i.e. I think WARN is particularly useful if it's something that, when a new release candidate goes out, any new warn lines (or drastic change in frequency of them) is something that is investigated before proceeding with the deploy.

in the past, I've also found warn to be useful for things that you _can somehow handle/recover_ but are _anyway unexpected_ (although, it's nice if that sort of thing goes to sentry)

Reply:
Makes sense! I've added both archetypes of `WARN`s to the description, and also mentioned your suggestion about investigating changes in frequency to `WARN` during the release process.